### PR TITLE
codeintel: LSIF validation

### PR DESCRIPTION
**Pull request by chrismwendt**
This adds a post-correlation validation step to LSIF upload processing. Here's what a validation error looks like:

CleanShot 2021-10-10 at 23 46 55@2x

Original motivation ([in Slack](https://sourcegraph.slack.com/archives/CHXHX7XAS/p1633905092280200?thread_ts=1633471921.198000&cid=CHXHX7XAS)) was a case where `item` edges from `implementationResult`s to ranges had incorrect `document` properties, which caused an infinite loop in the LSIF backend at query time.

### CODE REVIEW

**Comment by chrismwendt** [comment]
Found this old PR, @efritz anything else you'd like to see before merging?

File: lib/codeintel/lsif/conversion/correlate_test.go [hunk 1]
```diff
@@ -300,3 +300,16 @@ func TestCorrelateMetaDataRootX(t *testing.T) {
...
+
+func TestCorrelateConflictingDocumentProperties(t *testing.T) {
+	// dump4 is the same as dump1, but with one wrong "document" property
+	input, err := os.ReadFile("../testdata/dump4.lsif")
+	if err != nil {
+		t.Fatalf("unexpected error reading test file: %s", err)
+	}
+
+	_, err = correlateFromReader(context.Background(), bytes.NewReader(input), "")
+	if err == nil {
```
Comment (efritz) [inline lib/codeintel/lsif/conversion/correlate_test.go:312]: Should we actually test the error message here?

**Review by efritz — APPROVED** [review]

File: lib/codeintel/lsif/conversion/correlate.go [hunk 2]
```diff
@@ -374,6 +376,9 @@ func correlateContainsEdge(state *wrappedState, id int, edge Edge) error {
 		if _, ok := state.RangeData[inV]; !ok {
 			return malformedDump(id, inV, "range")
 		}
+		if doc, ok := state.rangeToDoc[inV]; ok && doc != edge.OutV {
+			return errors.Newf("validate: range %d is contained in document %d, but linked to a different document %d", inV, edge.OutV, doc)
+		}
 		state.Contains.AddID(edge.OutV, inV)
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply lib/codeintel/lsif/conversion/correlate.go:382]: There were conflicts with idset methods, but easily fixed.

File: lib/codeintel/lsif/conversion/correlate_test.go (hunk 1 above)
Comment (chrismwendt) [reply lib/codeintel/lsif/conversion/correlate_test.go:320]: Yep, added a check for the error.

**Pull request merged**